components/script/dom/bindings/str.rs

```diff
@@ -332,3 +346,93 @@ impl Extend<char> for DOMString {
         self.0.extend(iterable)
     }
 }
+
+/// https://html.spec.whatwg.org/multipage/#parse-a-month-string
+fn parse_month_string(value: &str) -> Result<(u32, u32), ()> {
+    // Step 1, 2, 3
+    let (year_int, month_int) = parse_month_component(value)?;
+
+    // Step 4
+    match value.split('-').nth(2) {
+        Some(_) => return Err(()),
+        None => {}
+    }
+    // Step 5
+    Ok((year_int, month_int))
+}
+
+/// https://html.spec.whatwg.org/multipage/#parse-a-date-string
+fn parse_date_string(value: &str) -> Result<(u32, u32, u32), ()> {
+    // Step 1, 2, 3
+    let (year_int, month_int, day_int) = parse_date_component(value)?;
+
+    // Step 4
+    match value.split('-').nth(3) {
+        Some(_) => return Err(()),
+        None => {}
+    }
+
+    // Step 5, 6
+    Ok((year_int, month_int, day_int))
+}
+
+/// https://html.spec.whatwg.org/multipage/#parse-a-month-component
+fn parse_month_component(value: &str) -> Result<(u32, u32), ()> {
+    // Step 3
+    let year = value.split('-').nth(0).ok_or(())?;
+    let month = value.split('-').nth(1).ok_or(())?;
+
+    // Step 1, 2
+    if year.len() < 4 ||
+        !year.chars().all(|c| c.is_digit(10)) ||
+        year == "0000" {
```

Comment:
Instead of checking whether every character is a digit, we can just parse the `year` into a `u32` above, and check whether the value is equal to 0.